### PR TITLE
Call onOpen on the reader thread and force writing on another thread.

### LINKED ISSUE #1667: WebSocket onOpen / onMessage race
`WebSocketCall` used to have the following for opening a web socket:
```java
    // Start a dedicated thread for reading the web socket.
    new Thread(new NamedRunnable("OkHttp WebSocket reader %s", request.urlString()) {
      @Override protected void execute() {
        while (webSocket.readMessage()) {
        }
      }
    }).start();

    listener.onOpen(webSocket, response);
```

This is racy in that a message could be read and delivered to the `onMessage` callback before `onOpen` had a chance to store a reference to the `WebSocket` in a field for a reply.

In the Autobahn testing, I naively made a change to the code:
```java
    listener.onOpen(webSocket, response);

    // Start a dedicated thread for reading the web socket.
    new Thread(new NamedRunnable("OkHttp WebSocket reader %s", request.urlString()) {
      @Override protected void execute() {
        //noinspection StatementWithEmptyBody
        while (webSocket.readMessage()) {
        }
      }
    }).start();
```

This fixes the race but presents a problem if `onOpen` does all of their interaction in a blocking fashion.

I have two ideas for rectifying this, and I am looking for more:
 * Expose an explicit call to start the reader thread in `onOpen` which would be called automatically after `onOpen` returned.
 * Start the reader thread on first interaction with the `WebSocket` instance or automatically after `onOpen` returned.

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1667.

### CODE REVIEW

File: samples/guide/src/main/java/com/squareup/okhttp/recipes/WebSocketEcho.java [hunk 1]
```diff
@@ -27,21 +27,28 @@ private void run() throws IOException {
...
   }
 
-  @Override public void onOpen(WebSocket webSocket, Response response)
-      throws IOException {
-    webSocket.sendMessage(TEXT, new Buffer().writeUtf8("Hello..."));
-    webSocket.sendMessage(TEXT, new Buffer().writeUtf8("...World!"));
-    webSocket.sendMessage(BINARY, new Buffer().writeInt(0xdeadbeef));
-    webSocket.close(1000, "Goodbye, World!");
+  @Override public void onOpen(final WebSocket webSocket, Response response) {
+    new Thread() {
```
Comment (swankjesse) [inline samples/guide/src/main/java/com/squareup/okhttp/recipes/WebSocketEcho.java]: might want to follow up with an executor service.

**Comment by swankjesse** [comment]
LGTM

**Comment by swankjesse** [comment]
```
[INFO] Starting audit...
/home/travis/build/square/okhttp/samples/guide/src/main/java/com/squareup/okhttp/recipes/WebSocketEcho.java:42:6: '}' is not followed by whitespace.
Audit done.
```

**Comment by JakeWharton** [comment]
Ugh. Gross.

**Pull request merged**